design/tracing-design.md

```diff
@@ -0,0 +1,170 @@
+- **Support custom tags to send additional span data**, including getting data from environment variables, request headers, etc.
+
+## Non Goals
+- **Per-HTTPProxy trace configuration.** There is a technical limitation to enabling per-proxy configuration. Trace settings are configured on the HTTP connection manager (HCM). Each TLS-enabled virtual host uses its own HCM, but *all* non-TLS virtual hosts share a single HCM. As such, while it's possible to have unique trace settings for each TLS virtual host, it's not possible to do the same for non-TLS virtual hosts.  Note that this same limitation has come up when designing external authorization and global rate limiting as well.
+- **Trace formats other than OpenTelemetry.** While Contour *may* choose to add direct support for other trace formats in the future, our hope is that by supporting the OpenTelemetry collector, users are able to export traces to the backend of their choice without having to directly configure it in Contour/Envoy.
+
+## High-Level Design
+The Contour configuration file will be extended with a new optional `tracing` section. This configuration block, if present, will enable tracing and will define the trace format and other properties needed to generate and export trace data.
+
+The `OpenTelemetry` trace format will be supported.
+
+The parameters defined in the configuration file will be used to configure each [HTTP connection manager's `tracing` section][3].
+
+Since the tracing backend is pluggable, Contour will not package any particular backend.
+The user is responsible for deploying and operating the tracing backend of their choice, and configuring Contour to make use of it.
+
+## Detailed Design
+
+### otel-collector ExtensionService
+
+When using Tracing,first,an ExtensionService must be defined for the otel-collector with the cluster-level details for the otel-collector itself.
+For example:
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
+metadata:
+  namespace: projectcontour
+  name: otel-collector
+spec:
+  protocol: h2c
+  services:
+    - name: otel-collector
+      port: 4317
+  timeoutPolicy:
+    response: 50ms
+```
+
+### Contour Configuration
+
+The `tracing` section of the Contour config file will look like:
+
+```yaml
+tracing:
+  # overallSampling defines the sampling rate of trace data.
+  overallSampling: 100
+  # maxPathTagLength defines maximum length of the request path to extract and include in the HttpUrl tag.
+  maxPathTagLength: 256
+  # customTags defines a list of custom tags with unique tag name.
+  # A customTag can only set one of literal, environment and requestHeaderName
+  customTags:
+    # tagName used to populate the tag name
+    - tagName: literal
+      # literal is a static custom tag value.
+      literal: 'this is literal'
+    - tagName: environment
```

Comment:
how much will users in the context of Contour/Envoy use this? is it common to set custom environment variables on the Envoy container?

Reply:
Obtaining the value of an environment variable is useful for some needs. For example, when there are multiple envoy instances, I can know which gateway generated the tracing through the HOST environment variable

Reply:
is the persona configuring tracing tags etc. going to be the same as the persona deploying Envoy? wondering if we should consider that these are likely not the same people in many cases so it doesn't quite make sense for them to have to have this cross-functional knowledge of deployment details

this might be something we can offer some syntactic sugar for or do by default for users rather than giving the generic ability to add tags based on environment variables

Reply:
Envoy itself supports obtaining the value of custom tags from environment variables. I think it is a good enhancement for us to provide this ability for users to use,and how to use it can be configured by the administrator.

Reply:
From some usecases, I think get tag from envs is good feature, for people who do the configurations for Contour/Envoy should have knowledge of this TracingConfiguration CRD, and most time for debug or issue tracking we need to know which Envoy container cause the problem, and if can't read tag from env, I think find pod name from trace is a hard work.

Reply:
providing the pod name/hostname in the trace labels makes a lot of sense, I'm just wondering if it is something we can provide automatically without exposing more surface area than we need to, in case there are secrets etc. in the environment

Reply:
one option would be to include the pod/hostname automatically for everyone, or offer a flag to enable it in the tag configuration e.g. `includePodDetails` that might add the name and namespace of the origin Envoy pod automatically, or even have a special tag type we can provide that is syntactic sugar for including origin envoy pod details

Reply:
this way also you don't have to redeploy Envoy if you want to modify what tags need to be set on the trace

Reply:
if the suggestion [here](https://github.com/projectcontour/contour/pull/4957/files#r1094639206) works and removing the `environment` tag value source, this LGTM

should add a note we would like to include the envoy pod name/namespace in the default tags

Reply:
we can also implement that specific feature as a follow-up

Reply:
This is a great suggestion, I'll update the design docs later